### PR TITLE
test: add factor option to adjust test timeouts

### LINKED ISSUE #18266: test: add option to functional test_runner to adjust test timeouts
It was [suggested](https://github.com/bitcoin/bitcoin/pull/18249#issuecomment-593944985) that we could have an additional argument for the [functional test_runner](https://github.com/bitcoin/bitcoin/blob/master/test/functional/test_runner.py) that modifies the timeouts by some factor. This would save having to randomly bump timeouts in the tests themselves, as users that require longer timeouts can opt in by `-factor=1.5` etc. 

Follow up to #18249.

#### Want to work on this issue?

The purpose of the `good first issue` label is to highlight which issues are suitable for a new contributor without a deep understanding of the codebase.

You do not need to request permission to start working on this. You are encouraged to comment on the issue if you are planning to work on it. This will help other contributors monitor which issues are actively being addressed and is also an effective way to request assistance if and when you need it.

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### LINKED ISSUE #18834: Intermittent failure in mempool_expiry "Unable to connect to bitcoind after 60s"
https://travis-ci.org/github/bitcoin/bitcoin/jobs/681265537#L3761

### PR DESCRIPTION
**Pull request by brakmic**
This PR adds a new option **factor** that can be used to adjust timeouts in various functional tests.
Several timeouts and functions from `authproxy`, `mininode`, `test_node` and `util` have been adapted to use this option. The factor-option definition is located in `test_framework.py`.

Fixes https://github.com/bitcoin/bitcoin/issues/18266
Also Fixes https://github.com/bitcoin/bitcoin/issues/18834

### CODE REVIEW

File: test/functional/test_framework/mininode.py [hunk 1]
```diff
@@ -585,7 +585,7 @@ def send_blocks_and_test(self, blocks, node, *, success=True, force_send=False,
                     self.send_message(msg_block(block=b))
             else:
                 self.send_message(msg_headers([CBlockHeader(block) for block in blocks]))
-                wait_until(lambda: blocks[-1].sha256 in self.getdata_requests, timeout=timeout, lock=mininode_lock)
+                wait_until(lambda: blocks[-1].sha256 in self.getdata_requests, timeout=timeout * self.options.factor, lock=mininode_lock)
```
Comment (MarcoFalke) [inline test/functional/test_framework/mininode.py:588]: Given that they are all based on `wait_until` wouldn't it make more sense to treat `wait_until` as the basic mechanism for timed waits and only multiply it in there?

Comment (brakmic) [reply test/functional/test_framework/mininode.py:588]: Meanwhile, I have changed a few other things, because I faced various p2p errors (the P2PInterface didn't have their factor variable, so I had to redo a few things). And yes, wait_until will be changed too.

**Comment by brakmic** [comment]
@MarcoFalke 

This one fails on travis, but succeeds on my macOS. Is there anything I could do to make it pass? 

```shell
mempool_reorg.py                                 | ✖ Failed  | 773 s
```

**Error Log**.

```shell
raise AssertionError("Mempool sync timed out:{}".format("".join("\n  {!r}".format(m) for m in pool)))
AssertionError: Mempool sync timed out:
```

File: test/functional/test_framework/mininode.py [hunk 2]
```diff
@@ -367,7 +367,7 @@ def on_version(self, message):
 
     def wait_for_disconnect(self, timeout=60):
         test_function = lambda: not self.is_connected
-        wait_until(test_function, timeout=timeout, lock=mininode_lock)
+        wait_until(test_function, timeout=timeout, lock=mininode_lock, factor=self.factor)
```
Comment (MarcoFalke) [inline test/functional/test_framework/mininode.py:370]: Given that they are all based on `def wait_until` wouldn't it make more sense to treat `wait_until` as the basic mechanism for timed waits and only multiply it in there?

File: test/functional/test_framework/mininode.py [hunk 3]
```diff
@@ -366,7 +366,7 @@ def on_version(self, message):
...
```
Comment (MarcoFalke) [inline test/functional/test_framework/mininode.py:369]:
```suggestion
        self.wait_until(test_function, timeout=timeout)
```

Could this be a member function, so that the lock and factor members don't have to be passed each time?

File: test/functional/test_framework/util.py [hunk 4]
```diff
@@ -208,9 +208,11 @@ def str_to_b64str(string):
 def satoshi_round(amount):
     return Decimal(amount).quantize(Decimal('0.00000001'), rounding=ROUND_DOWN)
 
-def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None):
+def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None, factor=1.0):
     if attempts == float('inf') and timeout == float('inf'):
-        timeout = 60
+        timeout = 60 * factor
+    else:
+        timeout = timeout * factor
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:215]:
```suggestion
    timeout = timeout * factor
```

can remove the else branch?

File: test/functional/test_framework/test_node.py [hunk 5]
```diff
@@ -324,13 +325,13 @@ def is_node_stopped(self):
...
     def wait_until_stopped(self, timeout=BITCOIND_PROC_WAIT_TIMEOUT):
-        wait_until(self.is_node_stopped, timeout=timeout)
+        wait_until(self.is_node_stopped, timeout=timeout, lock=None, factor=self.factor)
 
     @contextlib.contextmanager
     def assert_debug_log(self, expected_msgs, unexpected_msgs=None, timeout=2):
         if unexpected_msgs is None:
             unexpected_msgs = []
-        time_end = time.time() + timeout
+        time_end = time.time() + int(timeout * self.factor)
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_node.py:334]:
```suggestion
        time_end = time.time() + timeout * self.factor
```

why is this needed?

File: test/functional/test_framework/test_node.py [hunk 6]
```diff
@@ -76,7 +76,7 @@ def __init__(self, i, datadir, *, chain, rpchost, timewait, bitcoind, bitcoin_cl
         self.stderr_dir = os.path.join(self.datadir, "stderr")
         self.chain = chain
         self.rpchost = rpchost
-        self.rpc_timeout = timewait
+        self.rpc_timeout = int(timewait)
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_node.py:79]: why is this needed, when authproxy does the casting already?

File: test/functional/test_framework/test_node.py [hunk 7]
```diff
@@ -488,6 +489,7 @@ def add_p2p_connection(self, p2p_conn, *, wait_for_verack=True, **kwargs):
             kwargs['dstaddr'] = '127.0.0.1'
 
         p2p_conn.peer_connect(**kwargs, net=self.chain)()
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_node.py:491]:
```suggestion
        p2p_conn.peer_connect(**kwargs, net=self.chain, factor=self.factor)()
```

**Review by maflcko — APPROVED** [review]
ACK, just style nits

**Comment by MarcoFalke** [comment]
Could you please bump the factor to 2.0 for the valgrind run and mark this pull as fixing #18834 ?

**Comment by brakmic** [comment]
> Could you please bump the factor to 2.0 for the valgrind run and mark this pull as fixing #18834 ?

Yes, sure, will do it.

File: test/functional/test_framework/test_node.py (hunk 5 above)
Comment (brakmic) [reply test/functional/test_framework/test_node.py:334]: I had problems in test_node.py's *wait_for_rpc_connection*. More precisely, the polling for-loop that refused to iterate over float values. Hence the conversion to int. But if there is a way to make it work without conversion, I'll gladly try it out. Right now, I am going through your list of nits and adapting the code.

Comment (MarcoFalke) [reply test/functional/test_framework/test_node.py:334]: `time.time()` is a float, so I don't see how this can cause issues

File: test/functional/test_framework/test_node.py (hunk 7 above)
Comment (brakmic) [reply test/functional/test_framework/test_node.py:491]: In some previous variant of my code I encountered such a problem. Maybe just because I was still writing it. Right now, it seems that all those conversions aren't needed anymore. I'm just executing some tests and they're all passing.

File: test/functional/test_framework/test_node.py (hunk 6 above)
Comment (brakmic) [reply test/functional/test_framework/test_node.py:79]: In some previous variant of my code I encountered such a problem. Maybe just because I was still writing it. Right now, it seems that all those conversions aren't needed anymore. I'm just executing some tests and they're all passing.

File: test/functional/test_framework/test_node.py (hunk 7 above)
Comment (brakmic) [reply test/functional/test_framework/test_node.py:491]: I don't understand this change?

Comment (brakmic) [reply test/functional/test_framework/test_node.py:491]: Ah, yes, now I see it. It's being forwarded to derived class P2PInterface!

File: test/functional/test_framework/util.py [hunk 8]
```diff
@@ -208,9 +208,10 @@ def str_to_b64str(string):
 def satoshi_round(amount):
     return Decimal(amount).quantize(Decimal('0.00000001'), rounding=ROUND_DOWN)
 
-def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None):
+def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None, factor=2.0):
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:211]: Why increase the factor for everything? valgrind should be sufficient:

```diff
diff --git a/ci/test/00_setup_env_native_valgrind.sh b/ci/test/00_setup_env_native_valgrind.sh
index d5b38a1d88..c850c13bbe 100644
--- a/ci/test/00_setup_env_native_valgrind.sh
+++ b/ci/test/00_setup_env_native_valgrind.sh
@@ -13,7 +13,7 @@ export NO_DEPENDS=1
 if [[ "${TRAVIS}" == "true" && "${TRAVIS_REPO_SLUG}" != "bitcoin/bitcoin" ]]; then
   export TEST_RUNNER_EXTRA="wallet_disable"  # Only run wallet_disable as a smoke test to not hit the 50 min travis time limit
 else
-  export TEST_RUNNER_EXTRA="--exclude rpc_bind"  # Excluded for now, see https://github.com/bitcoin/bitcoin/issues/17765#issuecomment-602068547
+  export TEST_RUNNER_EXTRA="--exclude rpc_bind --factor=2"  # rpc_bind excluded for now, see https://github.com/bitcoin/bitcoin/issues/17765#issuecomment-602068547
 fi
 export GOAL="install"
 export BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --with-gui=no CC=clang CXX=clang++"  # TODO enable GUI

File: test/functional/test_framework/util.py [hunk 9]
```diff
@@ -208,9 +208,10 @@ def str_to_b64str(string):
...
     if attempts == float('inf') and timeout == float('inf'):
-        timeout = 60
+       timeout = 60 * factor
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:213]:
```suggestion
        timeout = 60
```

File: test/functional/test_framework/test_node.py [hunk 10]
```diff
@@ -324,13 +325,13 @@ def is_node_stopped(self):
         return True
 
     def wait_until_stopped(self, timeout=BITCOIND_PROC_WAIT_TIMEOUT):
-        wait_until(self.is_node_stopped, timeout=timeout)
+        wait_until(self.is_node_stopped, timeout=timeout, lock=None, factor=self.factor)
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_node.py:328]:
```suggestion
        wait_until(self.is_node_stopped, timeout=timeout, factor=self.factor)
```

None is the default value already

**Review by maflcko — APPROVED** [review]
ACK

File: test/functional/test_framework/util.py (hunk 8 above)
Comment (brakmic) [reply test/functional/test_framework/util.py:211]: True. Have adapted the scripts.

File: test/functional/test_framework/test_framework.py [hunk 11]
```diff
@@ -101,6 +101,7 @@ def __init__(self):
         self.bind_to_localhost_only = True
         self.set_test_params()
         self.parse_args()
+        self.rpc_timeout = int(self.rpc_timeout * self.options.factor) # optionally, increase timeout by a factor
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py:104]:
```suggestion
        self.rpc_timeout = self.rpc_timeout * self.options.factor # optionally, increase timeout by a factor
```

Is this still needed?

File: test/functional/test_framework/util.py [hunk 12]
```diff
@@ -208,9 +208,10 @@ def str_to_b64str(string):
 def satoshi_round(amount):
     return Decimal(amount).quantize(Decimal('0.00000001'), rounding=ROUND_DOWN)
 
-def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None):
+def wait_until(predicate, *, attempts=float('inf'), timeout=float('inf'), lock=None, factor=1.0):
     if attempts == float('inf') and timeout == float('inf'):
-        timeout = 60
+       timeout = 60
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:213]: why change the whitespace to 3 spaces? Seems unrelated

**Review by maflcko — APPROVED** [review]
ACK

File: test/functional/test_framework/test_framework.py (hunk 11 above)
Comment (brakmic) [reply test/functional/test_framework/test_framework.py:104]: I think it is. When I execute a test like this it throws the one error we talked about previously (*the range poll function*). 

```shell
./test/functional/p2p_tx_download.py                                                     (4s 89ms)  
2020-05-02T23:36:00.054000Z TestFramework (INFO): Initializing test directory /var/folders/7q/4ffytzk562dd2ky4bfg9_w7h0000gn/T/bitcoin_func_test_54bxj
085
2020-05-02T23:36:00.247000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/Users/brakmic/projects/bitcoin-dev/btc-factor/test/functional/test_framework/test_framework.py", line 452, in start_nodes
    node.wait_for_rpc_connection()
  File "/Users/brakmic/projects/bitcoin-dev/btc-factor/test/functional/test_framework/test_node.py", line 217, in wait_for_rpc_connection
    for _ in range(poll_per_s * self.rpc_timeout):
TypeError: 'float' object cannot be interpreted as an integer
```

File: test/functional/test_framework/util.py (hunk 12 above)
Comment (brakmic) [reply test/functional/test_framework/util.py]: Unintentional. Corrected. :)

**Comment by MarcoFalke** [comment]
Thanks! ACK 2742c3428633b6ceaab6714635dc3adb74bf121b
